### ISSUE TITLE
Custom syntax: "#1"

### ISSUE THREAD

**Issue by abesto** [rhaiscript/rhai#496]
I've been having a lot of fun trying to build a [LambdaMOO](https://www.sindome.org/moo-manual.html) clone with Rhai as the scripting language. (For some context: imagine a realtime multiplayer text-based game, where you can program the game from inside the game). I've tried something like this with Lua before, but Rhai has been *so much nicer* to work with!

Unfortunately I've hit a wall. One common operation in user code is referring to an object by its ID number; for example `#1`. AFAICT this is not currently possible to reproduce with Rhai:
* Custom unary operators are not supported
* `on_var` returns a constant, so (replacing `#` with `N` to get a valid variable name) `N1.name = "foobar"` fails ([code](https://github.com/abesto/roo/blob/7b6a57d1ccb06652e443d81d704e60a24ebb10a8/src/api.rs#L364-L372))
* Custom syntaxes seem to 1. not accept `#` as the first token, and also 2. require a space between tokens, so replacing `#` with `N`, we can only write `N 1` instead of `N1` ([code](https://github.com/abesto/roo/blob/7b6a57d1ccb06652e443d81d704e60a24ebb10a8/src/api.rs#L321-L340))

This is so tightly locked down, I suspect this may be a limitation of the underlying parser? Either way, would love any thoughts / ideas / guidance you can offer.

(Yes, I do realize the Rhai manual says that writing the core of a game in Rhai is a terrible idea, and that's exactly what I'm trying to do :P)

**Comment by schungx** [comment]
What you really want is a short-hand for ID-based access of a central object store.

Essentially `#1` maps to `objects[1]`.  And I suppose `objects` is an `Array`.

In that case, why not name it `N` and then use `N[1]`?  It is only two more characters than `#1` and it keeps to the Rhai syntax.

Alternatively you can run your Rhai scripts through a preprocessor to replace `#1`, `#2` ... with `N[...]`.

**Comment by abesto** [comment]
> What you really want is a short-hand for ID-based access of a central object store.

Correct. I'll also want other, slightly more complex shorthands (like e.g. `$property` -> `#0.property`) but the same solutions would work for both.

>  And I suppose objects is an Array.

No, but conceptually that's a good enough approximation. Feel free to check out the linked code if you want to see more.

> In that case, why not name it N and then use N[1]? It is only two more characters than #1 and it keeps to the Rhai syntax.

Several reasons:
* 2 characters more is 100% more characters for `#1`
* This is an experiment in remaining as close to the original LambdaMOO as possible; that does `#1`, so I want to do `#1`
* This is a hobby project, I don't have real requirements, and I feel like doing `#1` :D

> Alternatively you can run your Rhai scripts through a preprocessor to replace #1, #2 ... with N[...].

Not without grabbing the AST and mutating that. Consider a string that contains `#1`.

**Comment by schungx** [comment]
> This is a hobby project, I don't have real requirements, and I feel like doing `#1` :D

That's the right spirit!  Only do something when it is fun!

One thing to help is maybe to create a version of `on_var` which returns a mutable reference to any variable...

**Comment by abesto** [comment]
> One thing to help is maybe to create a version of on_var which returns a mutable reference to any variable...

Yeah that sounds reasonable, maybe. The interesting parts of the data structure are behind `Arc<RwLock<...>>` anyway because concurrency, so if I could convince Rust to allow an `on_var_mut` or something similar, then this could work. But also - the "cannot assign to constant" error is a Rhai runtime error, so there is more magic involved here than "just" the borrow checker. I'll give this a go anyway.

**Comment by abesto** [comment]
So umm, this is a bit awkward :D Not sure how I missed this last time, but the below actually works (up to a point, see further down in this comment):

```rust
    engine.register_custom_operator("#", 255).unwrap();
    engine.register_custom_syntax_raw(
        "#",
        |symbols, lookahead| match symbols.len() {
            1 if lookahead == "-" => Ok(Some("$symbol$".into())),
            1 => Ok(Some("$int$".into())),
            2 if symbols[1] == "-" => Ok(Some("$int$".into())),
            2 => Ok(None),
            3 => Ok(None),
            _ => unreachable!(),
        },
        false,
        move |_, inputs| {
            let id = if inputs.len() == 2 {
                assert_eq!(
                    inputs[0]
                        .get_literal_value::<rhai::ImmutableString>()
                        .unwrap(),
                    "-"
                );
                -inputs[1].get_literal_value::<ID>().unwrap()
            } else {
                inputs[0].get_literal_value::<ID>().unwrap()
            };
            Ok(Dynamic::from(O::new(id)))
        },
    );
```

This allows writing `#1` to reference the object with ID 1; and the trickery with `-` even allows negative IDs (yes, they're actually needed): `#-1` works just fine.

I'm now hitting a different issue though. It smells like, even with the "operator" precedence for `#` set to 255 (which is >240, the precedence of `.`), the custom syntax "binds" less tightly than `.`; or something else is going wrong (admittedly, I'm abusing the parser a fair bit here).

* `(#1).name` evaluates correctly to `"Wizard"`
* `#1.name` fails, saying: `Syntax error: Expecting ';' to terminate this statement (line 1, position 3)`

I also managed to use the same approach to implement the `$thing` syntax (which is equivalent to `(#0).thing`); that too suffers from the same problem. A quick experiment shows that the custom syntax will also need to explicitly know about assignment: `(#1).name = "foobar"` tells me `Syntax error: Expression cannot be assigned to`. Based on all this it occurs to me that what I *really* want is a custom syntax where the output is not a Rhai value, but instead a partial AST (better yet, a Rhai code snippet). As you noticed right away, what I *really* want here is a rewrite rule.

This remains a hobby project for me, so I don't want to take up too much of your time, but if all this triggers any insights, I'd love to read them.

Another potential approach would be enabling variable names to start with non-ASCII characters. I did a fair bit of diving into Rhai code at this point, but this is one I didn't look at deeply, so would also love your thoughts on that idea.

**Comment by schungx** [comment]
That's because `1.` is a floating point number! It is not a problem with precedence. But dots and indexing are bound extremely tightly...  I believe tighter than anything else.

Amazing that you got this to work! I myself don't know even know it is possible to register a custom syntax on a custom operator...

Once you get that working, I'd appreciate if you can do a simple example and I'll put that in the Book!

**Comment by schungx** [comment]
>  I really want is a custom syntax where the output is not a Rhai value, but instead a partial AST 

You're absolutely right. Something that takes a custom syntax at compile time but then translates to a code snippet that is compiled into the code stream.  Essentially recreating a C preprocessor.

**Comment by schungx** [comment]
In the process of testing out your solution, I've discovered a bug in Rhai that prevents a custom expression from being used in a binary expression.

For example `let x = 2 * #42;` will fail to parse.  This will be fixed.

**Comment by abesto** [comment]
> That's because 1. is a floating point number!

Lol, that's amazing. I suppose I could try to extend the custom syntax to match floats and act accordingly, but with all the other caveats listed above, I think I need to abandon this approach. I do feel much better now than before; now it's "here are these 5 edge-cases that will be super hard to get right", instead of just "I don't understand how any of this works".

> I myself don't know even know it is possible to register a custom syntax on a custom operator... Once you get that working, I'd appreciate if you can do a simple example and I'll put that in the Book!

Heh. I have the sense this has a bunch of weird edges, like I'm not sure it actually handles the precedence correctly (even if stuff like floats are not involved). But also, AFAICT this *only* works with `register_custom_syntax_raw`, so all the "here be dragons" warnings are in place already. I tried to come up with an example that smells useful and avoids (or at least doesn't directly encourage stepping into) the above pitfalls; here's the best I got, a weird alternative string syntax:

```rust
fn main() {
    let mut engine = Engine::new();

    engine
        .register_custom_operator("$", 255)
        .unwrap()
        .register_custom_syntax_raw(
            "$",
            |symbols, _lookahead| match symbols.len() {
                1 => Ok(Some("$ident$".into())),
                2 => Ok(None),
                _ => unreachable!(),
            },
            false,
            |_context, inputs| Ok(inputs[0].get_variable_name().unwrap().into()),
        );

    let result: String = engine.eval("$foobar").unwrap();
    println!("{}", result);
}
```

Output is, of course, `foobar`.

> Essentially recreating a C preprocessor.

Or Rust macros :P Though, even thinking about what it'd take to get something like those into Rhai makes my head hurt.

FWIW, for now I'll probably try string replacements, and try to manually filter out situations where they're *not* needed. That's probably strings, comments, and uhh... not much else. Can always add more heuristics, and other than maybe a few unneeded rewrites, it'll Just Work.

I think at this point this from my side this issue is ready to close.

**Comment by abesto** [comment]
P.S.: if you ever do get around to adding a preprocessor-y / macro-y thing, I'd love to test-drive it with this project.

**Comment by schungx** [comment]
> P.S.: if you ever do get around to adding a preprocessor-y / macro-y thing, I'd love to test-drive it with this project.

Actually there have been a few calls for it already.  There are actually C pre-processors out there that can simply be used, although I am not sure if they would recognize certain Rust features such as interpolated strings.  It would be much simpler to have a pre-processor that works with JavaScript, since Rust syntax most resembles it.

However, it is really not a high priority on my list, although probably a nice-to-have.  And most pre-processors won't let you start with a symbol (e.g. `#` or `$`) anyway!

**Comment by abesto** [comment]
> However, it is really not a high priority on my list

Yep, totally understandable.

> And most pre-processors won't let you start with a symbol (e.g. # or $) anyway!

... dammit :D

Hwell! This was an adventure :) I'll try a custom manual "preprocessor" next time I'm working on this project. Thank you for your time, and good luck with all things Rhai and non-Rhai!

**Comment by schungx** [comment]
> That's because `1.` is a floating point number!

Come to think of it... the tokenizer probably should look ahead to see if the period is followed by a letter, in which case it is most probably a method call...

**Comment by abesto** [comment]
FWIW I don't worry about tracking down the *specific* branch we hit in each "failure" case. To me it's pretty clear at this point that beyond just "this wasn't designed into Rhai", there _also_ isn't a set of small hacks that'd make this work well. Even getting this far exposes the application to a TON of Rhai implementation details, which, while cool, makes me feel uncomfortable :D

**Comment by schungx** [comment]
Actually I resolved it.  It _IS_ a bug.  With the fix, `#1.name` now works.

**Comment by abesto** [comment]
:O Oh no back down the rabbit hole I go

**Comment by schungx** [comment]
You can try pulling from https://github.com/schungx/rhai which should include the fix.

**Comment by abesto** [comment]
Can confirm this fixes all forms of property reads: https://gist.github.com/abesto/3566f3eb1c14c0891326af6b0efcddc7 (the `;` prefixes are just telling the game "this is code, not a normal game command")

Assignments remain a problem unfortunately.

**Comment by schungx** [comment]
> Assignments remain a problem unfortunately.

Yes, that one would be difficult to get around.

In any language, you have what you'd call an "L-Value" (I believe it originally came from C) meaning that anything that can be assigned to (thus the `L` for LHS).

In Rust, an l-value is basically a `&mut` reference to something that can be updated.  For example `x[i] = 42;` has `x[i]` as an l-value, which is a `&mut` reference to the i-th element of the array `x`.

Rhai has built-in support for computing l-values, but it would be extremely difficult to get a custom syntax expression to return one.  That's because there are currently tricks (akem... ok, hacks) to handle things like pointing to a particular character inside a string or a bit inside a number, which Rust doesn't support natively.

So, unless custom syntax is extended to produce l-values instead of simple values, assignments will not be simple.

However, it will be extremely simple to extend the custom syntax to include assignment statements.  All you need to do is to check the look-ahead to see if it is `=`, and if it is, parse another `$expr$`.

**Comment by abesto** [comment]
> However, it will be extremely simple to extend the custom syntax to include assignment statements. All you need to do is to check the look-ahead to see if it is =, and if it is, parse another $expr$.

It *would* be, if the custom syntax was handling property access. It's not. In `#0.test.test`, the custom syntax only handles `#0`; everything after is done by Rhai. In theory it *should* be possible to cover all interesting syntaxes that can follow an object reference (which ... should really *only* be property access, I think?) but at the same time doing that to an arbitrary depth with the current expressiveness of the raw custom syntax parser gets us back into "my head is starting to hurt" territory.

**Comment by schungx** [comment]
> if the custom syntax was handling property access. It's not.

Ah.  Gotcha.

**Comment by abesto** [comment]
FTR / for posterity: the code I used for this experiment (including a pinned `rhai` dependency) is at https://github.com/abesto/roo/commit/2f0ed05e3d7eb3d617621ff52ccf676a5f93507a
